src/database/queries/opportunities/fetchAll.sql

```diff
@@ -0,0 +1,34 @@
+SELECT o.id AS "opportunityId",
```

Comment:
I wonder if we might actually NOT want to select JOINs -- and instead leave it to the client.

At least in our first pass.

The client wanting a list of opportunities doesn't mean they necessarily care about the applications or canonical fields or schemas associated with those opportunities.  This is a lot of complexity (and DB work) that isn't needed for that use case.

With this approach we would start with a vanilla `GET /opportunity` that returns the flat opportunity list and a `GET /application?opportunityId=...` for the use case of "give me all applications for a given opportunity"

The downside of this is MAYBE more HTTP calls depending on the use case -- but at this point i'm not even usre we'll have use cases that benefit from a fully populated relationship map for Opportunity!

I think this approach might obviate some of the other concerns you were looking for feedback and dramatically simplify this endpoint.

What do you think?

Reply:
(Happy to talk through this by the way) -- just to underscore after re-reading this; I think this just-the-object-being-requested approach will simplify the other API endpoint too / we'll have a better shot at hitting our demo goal.

I also think it's expected in a RESTful api that I'm only getting the object I'm asking for (unless again the API supports asking for children as well but I'm not surprised when that kind of behavior requires a flag of some kind).

Reply:
Yes, in my first pass I had it as a pretty dumb endpoint, basically a pass-through to the `opportunity` entity/relation. It is certainly not an optimization to add the complexity of bringing along the application schema. The reason the application schema comes along here with opportunity is to satisfy the minimum-viable product having only these endpoints:

* Implement POST /applicants endpoint
* Implement GET /applicants endpoint
* Implement POST /applications endpoint
* Implement GET /applications endpoint
* Implement POST /opportunities endpoint
* Implement GET /opportunities/{opportunityId} endpoint
* Implement GET /opportunities endpoint
* Implement GET /canonicalFields endpoint

Given these endpoints, and given the DB schema, in order to post applications we need the application schema, therefore the application schema must be added in or inferred from one of three places:
* POST /applicants (doesn't make much sense here)
* POST /applications (probably makes some sense here)
* POST /opportunities (probably makes the most sense here)

Until we develop explicit `POST` and `GET /applicationSchemas` or unless we push the complexity into the `/applications` endpoint, we need to let the application schema be posted as part of the opportunity.

Note that applications do not come back with the current implementation you are looking at, only the latest application schema for the opportunity. In other words, if I wanted to submit an application for an `opportunity`, the call required is first `GET /opportunities` and then `POST /applications` (assuming the applicant has also already been posted).

The API presented to the caller does not necessarily have to match 1-1 with the underlying database schema (although it might be more straightforward to follow the API if, for example, you have some familiarity with the database schema for whatever reason).

The implementation here leaves the door open for the future possibility of either posting an opportunity with its application schema in one post or posting first a simple opportunity and separately posting an application schema, but for a minimum viable product, the application schema has to come from somewhere, so for now it would be sent with `POST /opportunities` and read with `GET /opportunities`.

Does this help?

Reply:
> having only these endpoints:

I'm not sure that's a requirement of the MVP right?  I think it'd be fine (maybe preferred / simpler?) to have GET and POST endpoints for `applicationForms`

Reply:
I guess what I'm getting at is that I think it would be a lot easier to say "we have a baseline API is dumb and very fast to implement" and then "let's add this complexity in a future PR"

Just from an MVP perspective I think we'd hit the finish line that way, whereas right now I don't think this PR is ready to merge and it introduces a ton of questions about how we want to handle joins in our schema queries, etc.

Reply:
I don't recall a reason, I think I made the guess from the list of issues here in GitHub, but we didn't make any determination on which ones would be included or excluded. I fully support simplifying the API and including `applicationForms` because it appears the extra data mashing is pretty darn slow (at least my feeble first-pass implementation was).

Reply:
> I guess what I'm getting at is that I think it would be a lot easier to say "we have a baseline API is dumb and very fast to implement and users CAN access all the data they want" and then "let's add some convenience / richer data responses in a future PR"

Yes, fully on board.

> Just from an MVP perspective I think we'd hit the finish line that way, whereas right now I don't think this PR is ready to merge and it introduces a ton of questions about how we want to handle joins in our schema queries, etc.

Yes, you highlighted another downside is that the data validation is no longer as simple as "take a response from the db and validate it" but is "take a response from the db, do a bunch of mashing, then validate it." So yeah, if we do the simpler pattern, I think we will be better off in the long run, let the complexity be transparent to the caller (by transparent I mean visible here, not invisible as we CS people sometimes backwardsly say).